File: ios/Classes/ExelbidPlugin.swift
import Flutter
import UIKit
import ExelBidSDK
import AdSupport
import AppTrackingTransparency

let METHOD_CHANNEL_ID = "exelbid_plugin"
let METHOD_CHANNEL_VIEW_ID = "exelbid_plugin/banner_ad"
let METHOD_CHANNEL_NATIVE_VIEW_ID = "exelbid_plugin/native_ad"
let METHOD_CHANNEL_MEDIATION_ID = "exelbid_plugin/mediation"

public class ExelbidPlugin: NSObject, FlutterPlugin {
    static var channel: FlutterMethodChannel?
    var messenger: FlutterBinaryMessenger
    var interstitial: EBInterstitialAdController?
    var videoManager: EBVideoManager?
    var mediations: [String : EBPMediation] = [:]

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        self.channel = FlutterMethodChannel(name: METHOD_CHANNEL_ID, binaryMessenger: registrar.messenger())
        let instance = ExelbidPlugin(messenger: registrar.messenger()) 

        if let channel = self.channel {
            registrar.addMethodCallDelegate(instance, channel: channel)
        }

        let bannerFactory = EBPBannerAdViewFactory(messenger: registrar.messenger())
        registrar.register(bannerFactory, withId: METHOD_CHANNEL_VIEW_ID)

        let nativeFactory = EBPNativeAdViewFactory(messenger: registrar.messenger())
        registrar.register(nativeFactory, withId: METHOD_CHANNEL_NATIVE_VIEW_ID)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
            case "requestTrackingAuthorization":
                if #available(iOS 14.0, *) {
                    ATTrackingManager.requestTrackingAuthorization { status in 
                        result(status.rawValue)
                    }
                } else {
                    // iOS 12이하 ATT 미적용 상태 처리 (3: Authorized, 2: Denied)
                    result(ASIdentifierManager.shared().isAdvertisingTrackingEnabled ? 3 : 2)
                }
            case "loadInterstitial": 
                if let arguments = call.arguments as? [String: Any], let adUnitId = arguments["ad_unit_id"] as? String {
                    let coppa = arguments["coppa"] as? Bool ?? false
                    let isTest = arguments["is_test"] as? Bool ?? false

                    self.loadInterstitial(adUnitId: adUnitId, coppa: coppa, isTest: isTest)
                }

                result(nil)
            case "showInterstitial":
                self.showInterstitial()

                result(nil)
            case "initMediation":
                if let arguments = call.arguments as? [String: Any],
                    let mediationUnitId = arguments["mediation_unit_id"] as? String,
                    let mediationTypes = arguments["mediation_types"] as? [String] {
                    self.mediations[mediationUnitId] = EBPMediation(unitId: mediationUnitId, types: mediationTypes, binaryMessenger: self.messenger)
                }
                result(nil)
            case "loadInterstitialVideo":
                if let arguments = call.arguments as? [String: Any], let adUnitId = arguments["ad_unit_id"] as? String {
                    let coppa = arguments["coppa"] as? Bool ?? false
                    let isTest = arguments["is_test"] as? Bool ?? false

                    self.loadInterstitialVideo(adUnitId: adUnitId, coppa: coppa, isTest: isTest)
                }

                result(nil)
            case "showInterstitialVideo":
                self.showInterstitialVideo()

                result(nil)
            default:
                result(FlutterMethodNotImplemented)
        }
    }

    func setInterstitialHandler() {

    }

    func loadInterstitial(adUnitId: String, coppa: Bool = false, isTest: Bool = false) {
        self.interstitial = EBInterstitialAdController(adUnitId: adUnitId)
        if let interstitial = self.interstitial {
            interstitial.delegate = self

            interstitial.coppa = "\(coppa ? 1 : 0)"
            interstitial.testing = isTest

            interstitial.loadAd()
        }
    }

    func showInterstitial() {
        self.interstitial?.showFromViewController()
    }
    
    func loadInterstitialVideo(adUnitId: String, coppa: Bool = false, isTest: Bool = false) {
        self.videoManager = EBVideoManager(identifier: adUnitId)
        
        if let videoManager = self.videoManager {
            // 광고의 효율을 높이기 위해 옵션 설정
            videoManager.coppa("\(coppa ? 1 : 0)")
            
            // 테스트 광고 설정 (true - 테스트 광고가 응답)
            videoManager.testing(isTest)

            videoManager.startWithCompletionHandler { (request, error) in
                if let error = error  {
                    print(">>> \(error.localizedDescription)")
                    ExelbidPlugin.channel?.invokeMethod("onVideoFailAd", arguments: ["error_message": error.localizedDescription])
                }else{
                   ExelbidPlugin.channel?.invokeMethod("onVideoLoadAd", arguments: nil)
                }
            }
        }
    }

    func showInterstitialVideo() {
        if let videoManager = videoManager, let uiViewController = topViewController {
            videoManager.presentAd(controller: uiViewController, delegate: self)
        } else {
            ExelbidPlugin.channel?.invokeMethod("onVideoFailAd", arguments: ["error_message": "Not Found ViewController"])
        }
    }
    
    var topViewController: UIViewController? {
        var mainWindow: UIWindow? = nil
        if #available(iOS 15, *) {
            // iOS 15 이상
            if let windowScene = UIApplication.shared.connectedScenes
                .compactMap({ $0 as? UIWindowScene })
                .first(where: { $0.activationState == .foregroundActive }) {
                mainWindow = windowScene.windows.first(where: { $0.isKeyWindow })
            }
        } else if #available(iOS 13, *) {
            // iOS 13 ~ iOS 14
            mainWindow = UIApplication.shared.windows.first(where: { $0.isKeyWindow })
        } else {
            // iOS 12 이하
            mainWindow = UIApplication.shared.keyWindow
        }
        
        var topViewController = mainWindow?.rootViewController
        
        if let presented = topViewController?.presentedViewController {
            topViewController = presented
        } else if let navController = topViewController as? UINavigationController {
            topViewController = navController.topViewController
        } else if let tabBarController = topViewController as? UITabBarController {
            topViewController = tabBarController.selectedViewController
        }
        
        return topViewController
    }
}

// MARK: - EBInterstitialAdControllerDelegate

extension ExelbidPlugin : EBInterstitialAdControllerDelegate{
    public func interstitialDidLoadAd(_ interstitial: EBInterstitialAdController?) {
        ExelbidPlugin.channel?.invokeMethod("onInterstitialLoadAd", arguments: nil)
    }

    public func interstitialDidFailToLoadAd(_ interstitial: EBInterstitialAdController?) {
        ExelbidPlugin.channel?.invokeMethod("onInterstitialFailAd", arguments: nil)
    }

    public func interstitialDidAppear(_ interstitial: EBInterstitialAdController?) {
        ExelbidPlugin.channel?.invokeMethod("onInterstitialShow", arguments: nil)
    }

    public func interstitialDidDisappear(_ interstitial: EBInterstitialAdController?) {
        ExelbidPlugin.channel?.invokeMethod("onInterstitialDismiss", arguments: nil)
    }

    public func interstitialDidReceiveTapEvent(_ interstitial: EBInterstitialAdController?) {
        ExelbidPlugin.channel?.invokeMethod("onInterstitialClickAd", arguments: nil)
    }
}


// MARK: - EBVideoDelegate

extension ExelbidPlugin : EBVideoDelegate {

    public func videoAdDidLoad(adUnitID: String) {
        ExelbidPlugin.channel?.invokeMethod("onVideoLoadAd", arguments: nil)
    }

    public func videoAdDidFailToLoad(adUnitID: String, error: any Error) {
        ExelbidPlugin.channel?.invokeMethod("onVideoFailAd", arguments: ["error_message": error.localizedDescription])
    }

    public func videoAdDidAppear(adUnitID: String) {
        ExelbidPlugin.channel?.invokeMethod("onVideoShow", arguments: nil)
    }

    public func videoAdDidDisappear(adUnitID: String) {
        ExelbidPlugin.channel?.invokeMethod("onVideoDismiss", arguments: nil)
    }

    public func videoAdDidReceiveTapEvent(adUnitID: String) {
        ExelbidPlugin.channel?.invokeMethod("onVideoClickAd", arguments: nil)
    }
}

class EBPMediation {
    var unitId: String
    var manager: EBMediationManager
    var channel: FlutterMethodChannel

    init(unitId: String, types: [String], binaryMessenger: FlutterBinaryMessenger) {
        self.unitId = unitId
        self.manager = EBMediationManager(adUnitId: unitId, mediationTypes: types)
        self.channel = FlutterMethodChannel(name: "\(METHOD_CHANNEL_MEDIATION_ID)_\(unitId)", binaryMessenger: binaryMessenger)
        self.channel.setMethodCallHandler { [weak self] (call, result) in
            guard let self = self else { return }
            if call.method == "loadMediation" {
                self.loadMediation(result)
            } else if call.method == "nextMediation" {
                self.nextMediation(result)
            }
        }

        channel.invokeMethod("onInitMediation", arguments: nil);
    }

    func loadMediation(_ result: @escaping FlutterResult) {
        self.manager.requestMediation { (manager, error) in
            DispatchQueue.main.async {
                if error != nil {
                    result(false)
                } else {
                    result(true)
                }
            }
        }
    }

    func nextMediation(_ result: @escaping FlutterResult) {
        if let mediation = manager.next() {
            result(["network_id": mediation.id, "unit_id": mediation.unit_id]);
        } else {
            result(nil)
        }
    }
}

File: android/src/main/kotlin/com/motivi/exelbid_plugin/ExelbidPlugin.kt
package com.motivi.exelbid_plugin

import android.app.Activity;
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.util.Pair
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.BinaryMessenger
import com.onnuridmc.exelbid.ExelBid
import com.onnuridmc.exelbid.ExelBidInterstitial
import com.onnuridmc.exelbid.common.ExelBidError
import com.onnuridmc.exelbid.common.OnInterstitialAdListener
import com.onnuridmc.exelbid.common.OnMediationOrderResultListener
import com.onnuridmc.exelbid.lib.ads.mediation.MediationOrderResult
import com.onnuridmc.exelbid.lib.ads.mediation.MediationType

const val METHOD_CHANNEL_ID = "exelbid_plugin/channel"
const val METHOD_CHANNEL_VIEW_ID = "exelbid_plugin/banner_ad"
const val METHOD_CHANNEL_NATIVE_VIEW_ID = "exelbid_plugin/native_ad"
const val METHOD_CHANNEL_MEDIATION_ID = "exelbid_plugin/mediation"

class ExelbidPlugin : FlutterPlugin, ActivityAware, MethodCallHandler {

    companion object {
        lateinit var channel : MethodChannel
    }

    private lateinit var context: Context
    private lateinit var messenger: BinaryMessenger
    private var activity: Activity? = null
    private var interstitial: ExelBidInterstitial? = null
    private var interstitialVideo: ExelBidInterstitial? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private var mediations: MutableMap<String, EBPMediation> = mutableMapOf()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        messenger = flutterPluginBinding.binaryMessenger

        channel = MethodChannel(messenger, METHOD_CHANNEL_ID)
        channel.setMethodCallHandler(this)

        context = flutterPluginBinding.applicationContext

        // Exelbid View Factory
        flutterPluginBinding.platformViewRegistry.registerViewFactory(METHOD_CHANNEL_VIEW_ID, EBPBannerAdViewFactory(messenger))
        flutterPluginBinding.platformViewRegistry.registerViewFactory(METHOD_CHANNEL_NATIVE_VIEW_ID, EBPNativeAdViewFactory(messenger))
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "requestTrackingAuthorization") {
            Thread {
                // 광고식별자 상태 처리  (3: Authorized, 2: Denied)
                try {
                    val adInfo = AdvertisingIdClient.getAdvertisingIdInfo(activity ?: context)
                    if (!adInfo.isLimitAdTrackingEnabled) {
                        result.success(3)
                    } else {
                        result.success(2)
                    }
                } catch (e: Exception) {
                    result.success(2)
                }
            }.start()
        } else if (call.method == "loadInterstitial") {
            val arguments = call.arguments as? Map<String?, Any?>
            if (arguments != null) {
                var adUnitId = arguments.get("ad_unit_id") as? String ?: ""
                var coppa = arguments.get("coppa") as? Boolean ?: true
                var isTest = arguments.get("is_test") as? Boolean ?: false

                loadInterstitial(adUnitId, coppa, isTest)                
            }

            result.success(null)
        } else if (call.method == "showInterstitial") {
            showInterstitial()
            result.success(null)
        } else if (call.method == "loadInterstitialVideo") {
            Log.d(javaClass.name, "[ExelbidPlugin] MethodChannel : loadInterstitialVideo")
            val arguments = call.arguments as? Map<String?, Any?>
            if (arguments != null) {
                var adUnitId = arguments.get("ad_unit_id") as? String ?: ""
                var timer = arguments.get("timer") as? Int ?: 0
                var coppa = arguments.get("coppa") as? Boolean ?: true
                var isTest = arguments.get("is_test") as? Boolean ?: false

                loadInterstitialVideo(adUnitId, timer, coppa, isTest)
            }

            result.success(null)
        } else if (call.method == "showInterstitialVideo") {
            showInterstitialVideo()
            result.success(null)
        } else if (call.method == "initMediation") {
            val arguments = call.arguments as? Map<*, *>
            if (arguments != null) {
                (arguments["mediation_unit_id"] as? String)?.let { unitId ->
                    (arguments["mediation_types"] as? List<String>)?.let { types ->
                        mediations[unitId] = EBPMediation(activity ?: context, unitId, types, messenger)
                    }
                }
            }

            result.success(null)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        mediations.clear()

        interstitial?.destroy()
        interstitial = null
        interstitialVideo?.destroy()
        interstitialVideo = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    private fun loadInterstitial(adUnitId: String, coppa: Boolean = false, isTest: Boolean = false) {
        interstitial?.destroy()

        val newInterstitial = ExelBidInterstitial(activity ?: context, adUnitId)
        interstitial = newInterstitial

        newInterstitial.setCoppa(coppa)
        newInterstitial.setTestMode(isTest)

        newInterstitial.setInterstitialAdListener(object : OnInterstitialAdListener {
            override fun onInterstitialLoaded() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoaded")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoaded post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoaded identity OK, channel=${channel}")
                    try {
                        channel.invokeMethod("onInterstitialLoadAd", null)
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoaded invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoaded invokeMethod Exception: $e")
                    } 
                }
            }

            override fun onInterstitialFailed(errorCode: ExelBidError?, statusCode: Int) {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailed (${errorCode?.errorCode}:$statusCode) : ${errorCode?.errorMessage}")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailed post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailed identity OK, channel=${channel}")
                    
                    try {
                        channel.invokeMethod("onInterstitialFailAd", mapOf("error_message" to errorCode?.errorMessage))
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailed invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailed invokeMethod Exception: $e")
                    }
                }
            }

            override fun onInterstitialShow() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShow")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShow post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShow identity OK, channel=${channel}")

                    try {
                        channel.invokeMethod("onInterstitialShow", null)
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShow invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShow invokeMethod Exception: $e")
                    }
                }
            }

            override fun onInterstitialDismiss() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismiss")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismiss post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismiss identity OK, channel=${channel}")
                    
                    try {
                        channel.invokeMethod("onInterstitialDismiss", null)
                        Log.d(javaClass.name, "[ExelbidPlugin]onInterstitialDismiss invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismiss invokeMethod Exception: $e")
                    }
                }
            }

            override fun onInterstitialClicked() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClicked")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClicked post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClicked identity OK, channel=${channel}")

                    try {
                        channel.invokeMethod("onInterstitialClickAd", null)
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClicked invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClicked invokeMethod Exception: $e")
                    }
                }
            }
        })

        newInterstitial.load()
    }

    private fun showInterstitial() {
        interstitial?.takeIf { it.isReady() }?.show()
    }

    private fun loadInterstitialVideo(adUnitId: String, timer: Int = 0, coppa: Boolean = false, isTest: Boolean = false) {
        interstitialVideo?.destroy()

        val newInterstitial = ExelBidInterstitial(activity ?: context, adUnitId)
        interstitialVideo = newInterstitial

        newInterstitial.setCoppa(coppa)
        newInterstitial.setTestMode(isTest)
        newInterstitial.setTimer(timer)

        newInterstitial.setInterstitialAdListener(object : OnInterstitialAdListener {
            override fun onInterstitialLoaded() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoadedVideo")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoadedVideo post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoadedVideo identity OK, channel=${channel}")
                    
                    try {
                        channel.invokeMethod("onVideoLoadAd", null)
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoadedVideo invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialLoadedVideo invokeMethod Exception: $e")
                    }
                }
            }

            override fun onInterstitialFailed(errorCode: ExelBidError?, statusCode: Int) {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailedVideo (${errorCode?.errorCode}:$statusCode) : ${errorCode?.errorMessage}")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailedVideo post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailedVideo identity OK, channel=${channel}")

                    try {
                        channel.invokeMethod("onVideoFailAd", mapOf("error_message" to errorCode?.errorMessage))
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailedVideo invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialFailedVideo invokeMethod Exception: $e")
                    }
                }
            }

            override fun onInterstitialShow() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShowVideo")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShowVideo post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShowVideo identity OK, channel=${channel}")

                    try {
                        channel.invokeMethod("onVideoShow", null)
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShowVideo invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialShowVideo invokeMethod Exception: $e")
                    }
                }
            }

            override fun onInterstitialDismiss() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismissVideo")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismissVideo post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismissVideo identity OK, channel=${channel}")

                    try {
                        channel.invokeMethod("onVideoDismiss", null)
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismissVideo invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialDismissVideo invokeMethod Exception: $e")
                    }
                }
            }

            override fun onInterstitialClicked() {
                Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClickedVideo")
                mainHandler.post {
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClickedVideo post executing : ${interstitial !== newInterstitial}")
                    if (interstitial !== newInterstitial) {
                        return@post
                    }
                    Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClickedVideo identity OK, channel=${channel}")

                    try {
                        channel.invokeMethod("onVideoClickAd", null)
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClickedVideo invokeMethod Success")
                    } catch(e: Exception) {
                        Log.d(javaClass.name, "[ExelbidPlugin] onInterstitialClickedVideo invokeMethod Exception: $e")
                    }
                }
            }
        })

        newInterstitial.loadAd()
    }

    private fun showInterstitialVideo() {
        interstitialVideo?.takeIf { it.isReady }?.show()
    }
}

class EBPMediation(
    private val context: Context,
    private val unitId: String,
    private val types: List<String>,
    binaryMessenger: BinaryMessenger
) {
    private val channel: MethodChannel
    private var mediationOrderResult: MediationOrderResult? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    init {
        channel = MethodChannel(binaryMessenger, "${METHOD_CHANNEL_MEDIATION_ID}_$unitId")
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "loadMediation" -> {
                    loadMediation(result)
                }
                "nextMediation" -> {
                    nextMediation(result)
                }
            }
        }

        mainHandler.post {
            channel.invokeMethod("onInitMediation", null);
        }
    }

    private fun loadMediation(result: Result) {
        ExelBid.getMediationData(context, unitId, convertToEnumList(types), object: OnMediationOrderResultListener {
            override fun onMediationOrderResult(orderResult: MediationOrderResult) {
                if(orderResult.getSize() > 0) {
                    mediationOrderResult = orderResult;
                    result.success(true)
                } else {
                    result.success(false)
                }
            }

            override fun onMediationFail(errorCode: Int, errorMsg: String) {
                result.success(false)
            }
        });
    }

    private fun nextMediation(result: Result) {
        val currentMediationPair: Pair<MediationType, String>? = mediationOrderResult?.poll()

        currentMediationPair?.let { pair ->
            val networkId: MediationType = pair.first
            val unitId: String = pair.second

            result.success(mapOf("network_id" to networkId.toString(), "unit_id" to unitId))
        } ?: result.success(null)
    }

    private fun convertToEnumList(types: List<String>): ArrayList<MediationType> {
        return types.mapNotNull { type -> MediationType.values().firstOrNull { it.toString() == type } }.toCollection(ArrayList())
    }
}
